kernel/condvar.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <toolchain.h>
+#include <ksched.h>
+#include <wait_q.h>
+
+static struct k_spinlock lock;
+
+int z_impl_k_condvar_init(struct k_condvar *condvar)
+{
+	z_waitq_init(&condvar->wait_q);
+	z_object_init(condvar);
+	return 0;
+}
+
+int z_impl_k_condvar_signal(struct k_condvar *condvar)
+{
+	int woken = 0;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	struct k_thread *thread = z_unpend_first_thread(&condvar->wait_q);
+
+	if (thread != NULL) {
+		arch_thread_return_value_set(thread, 0);
+		z_ready_thread(thread);
+		woken = 1;
+
+	}
+	z_reschedule(&lock, key);
+	return woken;
```

Comment:
What does the `woken` value mean here? According to documentation, this function returns 0 if ok, and any other value is error. According to this logic, if thread was woken, it is an error?